tamr_toolbox/utils/downstream.py

```diff
@@ -0,0 +1,245 @@
+import logging
+from tamr_unify_client.dataset.resource import Dataset
+from tamr_unify_client.project.resource import Project
+from tamr_unify_client.client import Client
+from typing import List, Optional, Dict
+from collections import OrderedDict
+from tamr_toolbox.models.project_type import ProjectType
+import re
+
+LOGGER = logging.getLogger(__name__)
+
+
+def datasets(dataset: Dataset, *, include_dependencies_by_name: bool = False) -> List[Dataset]:
+    """Returns a dataset's downstream datasets.
+
+    Args:
+        dataset: The target dataset.
+        include_dependencies_by_name: A Boolean indicating whether to return datasets with
+            dependencies suggested by name matching. Default to False.
+            No dependencies will be found by name if the `dataset` is not an unified dataset
+            either based on backened pipeline (if project still exists) or based on
+            regex (dataset name has suffix 'unified_dataset').
+
+    Returns:
+        List of Dataset objects ordered by number of its downstream dependencies.
+            Note that there can be bidirectional dependency so datasets with same number of
+            dependencies can depend on each other.
+    """
+    downstream_datasets = _find_downstream_datasets(
+        dataset.name,
+        include_dependencies_by_name=include_dependencies_by_name,
+        client=dataset.client,
+    )
+    # Remove the target dataset from the list
+    downstream_datasets.pop(dataset.name)
+    # Extract the list of the Dataset objects
+    downstream_datasets = list(downstream_datasets.values())
+    # sort list of downstream datasets by number of dependencies
+    # because parent dataset always has greater number of dependencies than its child datasets
+    # note that there can be bidirectional dependency so datasets with same number of
+    # dependencies can depend on each other
+    LOGGER.info("Sorting the downstream datasets by number of dependencies")
+    sorted_datasets = sorted(downstream_datasets, key=lambda x: _count_dependencies(x))
```

Comment:
You've defined a function for _count_dependencies but it is simple and only used once, I suggest simplifying to:

sorted_datasets = sorted(downstream_datasets, key=lambda x: len(x.usage().dependencies))